doc/webisoget.man
.TH WebISOGet 1

.SH NAME
WebISOGet - Get a web page from a WEBISO protected page.

.SH SYNTAX
\fBwebisoget\fP [\fIoptions\fP] 

.SH DESCRIPTION
\fBwebisoget\fP
Retrieves a web page via http or https, possibly following redirects 
and responding to forms.  Among the forms \fBwebisoget\fP will respond to
are those used by the \fBpubcookie\fP and \fBshibboleth\fP systems
during login.  In this way \fBwebisoget\fP can be used to retrieve web pages
protected by a webiso system such as \fBpubcookie\fP or \fBshibboleth\fP.
(see the web page for more up-to-date info.)


.SH OPTIONS
Options are processed as they are encountered.  Note that this makes
\fBwebisoget\fP's action very much dependent upon the order of options.
Normally one would specify one or more urls last.
.TP 12
.B -anchor \fIanchor_text\fP
Follow the link: \fC<a href=..>\fIanchor_text\fP\fC</a>\fP
.TP 12
.B -cache \fIfilename\fP
Cookies are retrieved from and saved to \fIfilename\fP.
\fBwebisoget\fP rewrites this file before exiting.
.TP 12
.B -cafile \fIfilename\fP
Specify a file of CA root certificates to verify SSL certificates.
.TP 12
.B -cert \fIfilename\fP
Specify a certificate to be used for client certificate authentication.
.TP 12
.B -certkey \fIfilename\fP
Specify the key for the client certificate.
.TP 12
.B -continue
Reprocess the last page
.TP 12
.B -cookies
Before exit, \fBwebisoget\fP will
report a list of cookies received during the session.
.TP 12
.B -debug
More output than \fIverbose\fP.
.TP 12
.B -debug2
More output than \fIdebug\fP.
.TP 12
.B -debug3
More output than \fIdebug2\fP.
.TP 12
.B -form \fIdescription\fP
Identifies a form to respond to.  (See FORM_DESCRIPTION)
.TP 12
.B -formfile \fIname\fP
Form information is in the indicated file. (See FORM_DESCRIPTION)
.TP 12
.B -frame \fIname\fP
Identifies a frame to follow. 
.TP 12
.B -map \fIalias=real\fP
Identifies a real host (\fIreal\fP) to be accessed instead 
of the alias name ((\fIalias\fP).  This may be used, for example,
to connect to a particular member of a cluster.
.TP 12
.B -maxhop \fIcount\fP
Maximum number of redirections to follow.
.TP 12
.B -maxtext \fIcount\fP
Maximum chars of text to stdout. Default=1M. Zero=unlimited.
.TP 12
.B -noverify 
Will not attempt to verify the certificate of the web server.
.TP 12
.B -out \fIfilename\fP
Write the last page to \fIfilename\fP.
.TP 12
.B -pagetimes 
Show elapsed time of operations.
.TP 12
.B -text
Displays the text of the pages received. If \fBwebisoget\fP follows
redirections, only the last page is displayed.
.TP 12
.B -timeout \fIseconds\fP
Time limit for page retrievals
.TP 12
.B -url \fIurl\fP
URL to retrieve.  \fBwebisoget\fP will retrieve this page and exit.
.TP 12
.B -verbose
Gives commentary on its progress.
.TP 12
.B -version
Show version and exit.

.SH INTERACTIVE MODE
If run without a URL specified, \fIwebisoget\fP will read
commands (options, without the hyphen) from standard input until end of file. 


.SH COOKIES
\fIwebisoget\fP receives and sends cookies.  If you want it to
save these between sessions, use the \fI-cache\fP option.
Specifying \fI-cookies\fP causes \fIwebisoget\fP to print
a report of all cookies received during the session.

.SH FORM_DESCRIPTION
\fIwebisoget\fP will fill out and submit forms.
A form description contains the form name, possibly a domain name,
and the name and value  pairs for some form inputs. The values are in a syntax
similar to that used by web forms.  The fields
are separated by semi-colons.  Specify \fCname=;\fP to match
a form with no name.  Omit the name to match any form.  

If a domain name (\fCdomain=\fIsome.domain.org;\fP) is specified only
forms found on that domain will match.

If a submit name (\fCsubmit_name=\fIname;\fP) is specified only the
form with a submit button by that name will match.

If a submit value (\fCsubmit_value=\fIvalue;\fP) is specified only the
form with a submit button with that value will match.

For example, this form will achieve a pubcookie login
as user "spud" with password "potato".

.I name=query; user=spud; pass=potato;

A formfile is a file with one form description per line.

.SH LIBRARY API
The functions of \fIwebisoget\fP are available as a c API.
There is no additional documentation for this, but the
main program source shows quite clearly how to use the library.
Specify at least \fC-lwebisoget -lcurl -lssl -lcrypto -lz -lcurl\fP
to your linker.


.SH COPYRIGHT

.nf
  ========================================================================
  Copyright (c) 2004-2006 The University of Washington

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
  ========================================================================

.fi


.SH AUTHOR
Jim Fox, University of Washington, fox@washington.edu



CHANGES
Vers 2.7

??
Force ssl connections to be SSLv3


Vers 2.6

urldecode any incoming pre-encoded form text
(see encode_form_text)


Vers 2.4

remove local definition of strndup
change curl/types.h to curl/curl/typecheck-gcc.h


Vers 2.02

1) Translate '&#....;' to ascii char

Vers 2.01

1) Fix show-page at end of sequence
2) [2/27/2009] add webisoget.spec for rpmbuild

Vers 2.00

1) Refactoring and fixing of cookie handling
2) Refactoring and fixing of form handling

Vers 1.49

1) Fix poor cookie handling
2) Add post/put/delete

Version 1.47

1) Allow multiple, comma-separated cookies per header

Version 1.46

1) Fix something?

Version 1.45

1) Allow '\;' and '\n' in form text for semicolon and newline.

2) Form response includes only one "submit" button.
   If user specifies one it is used, else just the first encountered.

3) Textareas in forms are processed better.

4) Multiple values can be specified in form descriptions.
   e.g. item=aaa; item=bbb; ...

5) "-noverify" option works.

6) More use of strcasecmp.

7) Better parsing of input key=value strings.




webisoget.c
/* ========================================================================
 * Copyright (c) 2004-2006 The University of Washington
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ========================================================================
 */


/* Retrieve a web page, following redirections, and 
   responding to some forms.

   This allows us to obtain , for example, pubcookies and thereby
   retrieve pubcookie protected pages.

*/

#include "webisoget.h"

char *postdata = NULL;
FILE *postfile = NULL;
char *putdata = NULL;
FILE *putfile = NULL;

int interactive = 0;
extern int delete_op;

static char *user_agent = USER_AGENT_GECKO;

static char *prog;
struct {
 char *cmd;
 char *help;
} commands[] = {
 {"anchor anchor_text ", "<a...> link to follow"},
 {"cache file_name    ", "cookie cache for restore/save"},
 {"cafile file_name   ", "CA certificates (PEM)"},
 {"cert file_name     ", "Client certificate (PEM)"},
 {"key file_name      ", "Client certificate key (PEM) [aka 'certkey']"},
 {"continue           ", "reprocess last page"},
 {"cookies            ", "enable cookie report"},
 {"debug              ", "for more commentary"},
 {"debug2             ", "for more commentary"},
 {"debug3             ", "for even more commentary "},
 {"delete             ", "next request will be DELETE"},
 {"header header      ", "add header "},
 {"form form_info     ", "form info"},
 {"formfile file_name ", "form info in file"},
 {"frame frame_name   ", "frame to follow"},
 {"map virt=real      ", "mapping of hostnames"},
 {"maxhop count       ", "max number of redirects to follow (def=20)"},
 {"maxtext count      ", "max chars of text to stdout (def=1M, 0=unlimited)"},
 {"out output_file    ", "write final page to output_file"},
 {"pagetimes          ", "show elapsed time per page"},
 {"postfile file      ", "file to post"},
 {"putfile file       ", "file to put"},
 {"quit               ", "exit"},
 {"text               ", "enable display of page text"},
 {"timeout seconds    ", "set timeout in seconds (default=60)"},
 {"url url            ", "url to get"},
 {"verbose            ", "for commentary"},
 {"version            ", "show version"},
 { NULL, NULL }
};
 
 
static void usage()
{
   int c;
   if (interactive) {
     printf("Commands..\n");
     for (c=0; commands[c].cmd; c++) {
        printf("   %s %s\n", commands[c].cmd, commands[c].help);
     }
     printf("\n");
   } else {
     fprintf(stderr, "usage: %s [commands] \n", prog);
     for (c=0; commands[c].cmd; c++) {
        fprintf(stderr, "   -%s %s\n", commands[c].cmd, commands[c].help);
     }
     fprintf(stderr,"Commands are processed in the order sepcified.\n");
     fprintf(stderr,"Absence of a url enters stdin command mode.\n");
     fprintf(stderr, "\n");
     exit(1);
   }
}

static void helper()
{
   int c;
   for (c=0; commands[c].cmd; c++) {
      fprintf(stderr, "   %s %s\n", commands[c].cmd, commands[c].help);
   }
   fprintf(stderr, "\n");
}

/* read from a file into a string */

static char *get_file(char *file)
{
   FILE *f = fopen(file,"r");
   char *str;
   long l;

   if (!f) return (NULL);
   fseek(f, 0, SEEK_END);
   l = ftell(f);
   rewind(f);

   str = (char*) malloc(l+1);
   fread(str, l, 1, f);
   fclose(f);

   return(str);
}



static void set_debug(WebGet W, int d)
{
   W->debug = d;
   if (d>1) curl_easy_setopt(W->curl, CURLOPT_VERBOSE, 1);
}

/* process a command */

static WebPage current_page = NULL;
static int didurl = 0;

static int docmd(WebGet W, char *cmd, char *arg)
{
   char                inbuf[1024];
   char               *p;
   int                 headers = 1;

   char *url = NULL;
   WebPage page;
   char *s;
   char curl_errtxt[CURL_ERROR_SIZE];

   curl_easy_setopt(W->curl, CURLOPT_SSL_VERIFYPEER, 0);
   curl_easy_setopt(W->curl, CURLOPT_SSL_VERIFYHOST, 0);
   curl_easy_setopt(W->curl, CURLOPT_ERRORBUFFER, curl_errtxt);

   /* parameter setting */
   if (!strcmp(cmd,"verbose")) {
      set_debug(W, 1);
      curl_easy_setopt(W->curl, CURLOPT_VERBOSE, 1);
   } else if (!strcmp(cmd,"debug")) {
      set_debug(W, 2);
   } else if (!strcmp(cmd,"debug2")) {
      set_debug(W, 2);
   } else if (!strcmp(cmd,"debug3")) {
      set_debug(W, 3);

   } else if (!strcmp(cmd,"cookies")) {
      W->show_cookies = 1;

   } else if (!strcmp(cmd,"pubcookies")) {
      /* obsolete option */
      W->show_cookies = 1;

   } else if (!strcmp(cmd,"text")) {
      W->show_text = 1;

   } else if (!strcmp(cmd,"maxhop")) {
      if (!arg) usage();
      W->maxhop = atoi(arg);
      if (W->maxhop<0) usage();

   } else if (!strcmp(cmd,"maxtext")) {
      if (!arg) usage();
      W->maxxbuf = atoi(arg);
      if (W->maxxbuf<0) usage();

   } else if (!strcmp(cmd,"timeout")) {
      if (!arg) usage();
      W->timeout = atoi(arg);
      if (W->timeout<=0) usage();
      curl_easy_setopt(W->curl, CURLOPT_TIMEOUT, W->timeout);

   } else if (!strcmp(cmd,"cache")) {
      if (!arg) usage();
      W->cache_name = strdup(arg);
      restore_cookies(W);

   } else if (!strcmp(cmd,"form")) {
      if (!arg) usage();
      load_known_form(W, arg);

   } else if (!strcmp(cmd,"formfile")) {
      if (!arg) usage();
      load_known_form_from_file(W, arg);

   } else if (!strcmp(cmd,"frame")) {
      if (!arg) usage();
      add_frame(W, arg);

   } else if (!strcmp(cmd,"anchor")) {
      if (!arg) usage();
      add_anchor(W, arg);

   } else if (!strcmp(cmd,"map")) {
      if (!arg) usage();
      add_host_map(W, arg);

   } else if (!strcmp(cmd,"postdata")) {
      if (!arg) usage();
      postdata = strdup(arg);

   } else if (!strcmp(cmd,"postfile")) {
      if (!arg) usage();
      postdata = get_file(arg);

   } else if (!strcmp(cmd,"putdata")) {
      if (!arg) usage();
      putdata = strdup(arg);

   } else if (!strcmp(cmd,"putfile")) {
      if (!arg) usage();
      putfile = fopen(arg, "r");

   } else if (!strcmp(cmd,"header")) {
      if (!arg) usage();
      add_header(W, arg);

   } else if (!strcmp(cmd,"url")) {
      if (!arg) usage();
      if (current_page) {
         free_page(current_page);
         current_page = NULL;
      }
      page = get_one_page(W, arg, NULL);
      delete_op = 0;
      if (page) page = process_pages(page);
      if (!page) {
          if (W->curl_err) fprintf(stderr, "Curl error %d: %s\n", W->curl_err, curl_errtxt);
          return (0);
      }
      didurl++;
      current_page = page;

   } else if ((!strcmp(cmd,"continue"))||(!strcmp(cmd,"c"))) {
      if (current_page) current_page = process_pages(current_page);

   } else if ((!strcmp(cmd,"bin")) || (!strcmp(cmd,"out"))) {
      if (!arg) usage();
      W->binfile = fopen(arg,"w");
      if (!W->binfile) {
         perror("binfile");
         if (!interactive) exit (1);
      }

   } else if (!strcmp(cmd,"delete")) {
      delete_op = 1;

   } else if (!strcmp(cmd,"cert")) {
      if (!arg) usage();
      curl_easy_setopt(W->curl, CURLOPT_SSLCERT, strdup(arg));
      curl_easy_setopt(W->curl, CURLOPT_SSLCERTTYPE, "PEM");

   } else if ((!strcmp(cmd,"certkey")) || (!strcmp(cmd, "key"))) {
      if (!arg) usage();
      curl_easy_setopt(W->curl, CURLOPT_SSLKEY, strdup(arg));
      curl_easy_setopt(W->curl, CURLOPT_SSLKEYTYPE, "PEM");

   } else if (!strcmp(cmd,"cafile")) {
      if (!arg) usage();
      curl_easy_setopt(W->curl, CURLOPT_CAINFO, strdup(arg));
      curl_easy_setopt(W->curl, CURLOPT_SSL_VERIFYPEER, 1);
      curl_easy_setopt(W->curl, CURLOPT_SSL_VERIFYHOST, 0);
      verify_peer = 1;

   } else if (!strcmp(cmd,"noverify")) {
      curl_easy_setopt(W->curl, CURLOPT_SSL_VERIFYPEER, 0);
      curl_easy_setopt(W->curl, CURLOPT_SSL_VERIFYHOST, 0);
      verify_peer = 0;

   } else if (!strcmp(cmd,"pagetimes")) {
      W->show_pt = 1;

   } else if (!strcmp(cmd,"agent")) {
      if (!arg) usage();
      W->user_agent = strdup(arg);

   } else if (!strcmp(cmd,"version")) {
      fprintf(stdout, "%s version %s\n", PACKAGE_NAME, PACKAGE_VERSION);
      exit (0);

   } else if (!strcmp(cmd,"quit")) {
      save_cookies(W);
      exit (0);

   } else usage();

   return (1);
}


int  main(int argc, char *argv[])
{
   char inbuf[1024];
   char *p;
   int  headers = 1;
   int  r;
   WebGet W;

   char *cmd, *arg;

   prog = (argv++)[0];
   tzset();

   /* Initialize library */

   W = new_WebISOGet();

   /* process command line commands */

   while (--argc) {

      cmd = (argv++)[0]+1;
      if ((argc>1) && (*(argv[0])!='-')) {
         arg = (argv++)[0];
         argc--;
      } else arg = NULL;
      
      if (!docmd(W, cmd, arg)) exit(1);

   }

   /* no url on command line means get instructions from stdin */

   if (!didurl) {

      interactive = 1;
      for (;;) {
       fputs("cmd: ", stderr);
       if (fgets(inbuf,1024, stdin)) {
         char *n;
         char *cmd, *arg;
         if (!*inbuf) break;
         if (n=strchr(inbuf,'\n')) *n = '\0';

         /* get command and arg */
         cmd = inbuf;
         while (*cmd==' ') cmd++;  /* delete leading spaces */
         if (!*cmd) continue;
         arg = cmd;
         while (isalnum(*arg)) arg++;
         *arg++ = '\0';
         while (*arg==' ') arg++;
         if (*arg) {
           n = arg + strlen(arg) - 1;
           while (isspace(*n)) *n-- = '\0'; /* delete trailing spaces */
         }
         
         if (!docmd(W, cmd, arg)) exit (1);
       } else {
         puts("");
         break;
       }
      }
   }

   /* save the received cookies if all went OK */

   save_cookies(W);
   exit (0);
}

